File: packages/core-php/src/AuthData.php
<?php

declare(strict_types=1);

namespace OpenPasskey;

class AuthData
{
    public readonly string $rpIdHash;
    public readonly int $flags;
    public readonly int $signCount;
    public readonly ?string $credentialId;
    public readonly ?string $credentialKey;

    private const MIN_LEN = 37;

    public static function parse(string $authData, bool $expectCredData): self
    {
        if (strlen($authData) < self::MIN_LEN) {
            throw new WebAuthnException('authenticator_data_too_short');
        }

        $instance = new self();
        $instance->rpIdHash = substr($authData, 0, 32);
        $instance->flags = ord($authData[32]);
        $instance->signCount = unpack('N', substr($authData, 33, 4))[1];

        $hasAttestedCredData = ($instance->flags & 0x40) !== 0;

        if ($expectCredData) {
            if (!$hasAttestedCredData) {
                throw new WebAuthnException('no_attested_credential_data');
            }
            $rest = substr($authData, 37);
            if (strlen($rest) < 18) {
                throw new WebAuthnException('authenticator_data_too_short');
            }
            $credIdLen = unpack('n', substr($rest, 16, 2))[1];
            $rest = substr($rest, 18);
            if (strlen($rest) < $credIdLen) {
                throw new WebAuthnException('authenticator_data_too_short');
            }
            $instance->credentialId = substr($rest, 0, $credIdLen);
            $instance->credentialKey = substr($rest, $credIdLen);
        } else {
            $instance->credentialId = null;
            $instance->credentialKey = null;
        }

        return $instance;
    }

    public static function verifyRpIdHash(string $authDataRpIdHash, string $rpId): void
    {
        $expected = hash('sha256', $rpId, true);
        if (!hash_equals($expected, $authDataRpIdHash)) {
            throw new WebAuthnException('rp_id_mismatch');
        }
    }

    public static function checkFlags(int $flags, bool $requireUserVerification): void
    {
        if (($flags & 0x01) === 0) {
            throw new WebAuthnException('user_presence_required');
        }
        if ($requireUserVerification && ($flags & 0x04) === 0) {
            throw new WebAuthnException('user_verification_required');
        }
        // BS must be 0 if BE is 0 (§6.3.3)
        if (($flags & 0x08) === 0 && ($flags & 0x10) !== 0) {
            throw new WebAuthnException('invalid_backup_state');
        }
    }

    private function __construct()
    {
    }
}

File: packages/core-ruby/lib/open_passkey/auth_data.rb
# frozen_string_literal: true

module OpenPasskey
  class AuthData
    attr_reader :rp_id_hash, :flags, :sign_count, :credential_id, :credential_key

    MIN_LEN = 37

    def self.parse(auth_data, expect_cred_data)
      auth_data = auth_data.b if auth_data.encoding != Encoding::ASCII_8BIT

      if auth_data.bytesize < MIN_LEN
        raise WebAuthnError.new("authenticator_data_too_short")
      end

      instance = new
      instance.instance_variable_set(:@rp_id_hash, auth_data.byteslice(0, 32))
      instance.instance_variable_set(:@flags, auth_data.getbyte(32))
      instance.instance_variable_set(:@sign_count, auth_data.byteslice(33, 4).unpack1("N"))

      has_attested_cred_data = (instance.flags & 0x40) != 0

      if expect_cred_data
        unless has_attested_cred_data
          raise WebAuthnError.new("no_attested_credential_data")
        end
        rest = auth_data.byteslice(37..)
        if rest.bytesize < 18
          raise WebAuthnError.new("authenticator_data_too_short")
        end
        cred_id_len = rest.byteslice(16, 2).unpack1("n")
        rest = rest.byteslice(18..)
        if rest.bytesize < cred_id_len
          raise WebAuthnError.new("authenticator_data_too_short")
        end
        instance.instance_variable_set(:@credential_id, rest.byteslice(0, cred_id_len))
        instance.instance_variable_set(:@credential_key, rest.byteslice(cred_id_len..))
      else
        instance.instance_variable_set(:@credential_id, nil)
        instance.instance_variable_set(:@credential_key, nil)
      end

      instance
    end

    def self.verify_rp_id_hash(auth_data_rp_id_hash, rp_id)
      expected = OpenSSL::Digest::SHA256.digest(rp_id)
      unless secure_compare(expected, auth_data_rp_id_hash)
        raise WebAuthnError.new("rp_id_mismatch")
      end
    end

    def self.check_flags(flags, require_user_verification)
      if (flags & 0x01) == 0
        raise WebAuthnError.new("user_presence_required")
      end
      if require_user_verification && (flags & 0x04) == 0
        raise WebAuthnError.new("user_verification_required")
      end
      if (flags & 0x08) == 0 && (flags & 0x10) != 0
        raise WebAuthnError.new("invalid_backup_state")
      end
    end

    private_class_method :new

    def self.secure_compare(a, b)
      return false unless a.bytesize == b.bytesize
      OpenSSL.fixed_length_secure_compare(a, b)
    end
  end
end
